keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+- The upstream eviction API does not currently support evictions in background
+  implemented by external components. For that, there's no community
+  dedicated error code nor a response text for this type of eviction.
+- The eviction in background error code will be temporary mapped to 429 (TooManyRequests).
+- The response text will be searched for strings
+  containing "Eviction triggered" prefix or similar.
+- The descheduler will still treat all the pods with eviction in background as cattle.
+  It will only prioritize these pods to be processed sooner.
+  There may still be unnecessary/sub-optimal evictions.
+
+### Risks and Mitigations
+
+- If an external component responsible for eviction in background
+  does not work properly the descheduler might evict less pods than expected.
+  Leaving pods that require eviction (e.g. long running or violating third
+  party policies) running longer than expected.
+  The descheduler will need to implement a TTL mechanism to abandon
+  eviction requests that take longer than expected.
+- Most of the heavy lifting happens in the external components.
+  The descheduler only acknowledges pods that are evicted in background.
+  All the internal counters take these pods into account.
+  When these pods are not properly evicted the internal counters might limit
+  eviction of other pods that (when evicted) might improve overall cluster
+  resource utilization.
+- This new functionality needs to stay feature gated until
+  the evacuation API is available and stable.
+
+## Design Details
+
+### Implementation Strategies
+
+#### API changes
+
+- **New category of pods for eviction**: Two new annotations are introduced:
+  - `descheduler.alpha.kubernetes.io/request-evict-only`: to signal pods whose
+    eviction is not expected to be completed right away.
+    Instead, an eviction request is expected to be intercepted by an external
+    component which will initiate the eviction process for the pod.
+    E.g. by performing live migration or scheduling the eviction to a more suitable time.
+  - `descheduler.alpha.kubernetes.io/eviction-in-progress`: to mark pods whose
+    eviction was initiated by an external component.
+- **New eviction limits**: introduction of new `MaxNoOfPodsEvacuationsPerNode` and
+  `MaxNoOfPodsEvacuationsPerNamespace` feature gated v1alpha1 fields.
+- **New configurable list of eviction message prefixes (introduced later when needed)**:
+  each third party validating admission webhook can provide different response message.
+  It is not feasible to ask every party to change their messages for a temporary solution
+  until the evacuation API is in place. Thus, temporary introducing a new feature
+  gated `evictionMessagePrefixes` field (`[]string`) for a list of known prefixes
+  signaling an eviction in background.
+  Normally, "Eviction triggered" prefix will be acknowledged by default.
```

Comment:
how is this going to change with the Evacuation API?

Reply:
Deprecated and removed eventually.